x/meicdp/handler.go

```diff
@@ -117,3 +180,259 @@ func handleSetSourceChannel(ctx sdk.Context, keeper Keeper, msg types.MsgSetSour
+		)
+	}
+
+	msg, err := keeper.GetMsg(ctx, msgID)
+	if err != nil {
+		return sdkerrors.Wrapf(
+			types.ErrMsgNotFound,
+			"cannot get stored message",
+		)
+	}
+
+	switch msg := msg.(type) {
+	case types.MsgUnlockCollateral:
+		err := handleMsgUnlockCollateral(ctx, keeper, msg, collateralPrice)
+		if err != nil {
+			return err
+		}
+
+	case types.MsgBorrowDebt:
+		err := handleMsgBorrowDebt(ctx, keeper, msg, collateralPrice)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+// handleMsgUnlockCollateral handles the unlock collateral message after receives oracle packet
+func handleMsgUnlockCollateral(ctx sdk.Context, keeper Keeper, msg MsgUnlockCollateral, collateralPrice uint64) error {
+
+	cosmosHubChannelID, err := keeper.GetChannel(ctx, CosmosHubChain, "transfer")
+	denom := fmt.Sprintf("transfer/%s/%s", cosmosHubChannelID, types.AtomUnit)
+
+	cdp := keeper.GetCDP(ctx, msg.Sender)
+
+	unlockAmount := sdk.NewCoin(denom, sdk.NewInt(int64(msg.Amount)))
+	unlockAmountCoins := sdk.NewCoins(unlockAmount)
+
+	// Subtract collateral on CDP
+	unlockAmountInt := new(big.Int).SetUint64(msg.Amount)
+	collateralAmount := new(big.Int).SetUint64(cdp.CollateralAmount)
+	collateralAmount.Sub(collateralAmount, unlockAmountInt)
+
+	minimumCollateralAmount := new(big.Int).SetUint64(0)
+	if collateralAmount.Cmp(minimumCollateralAmount) == -1 {
+		return sdkerrors.Wrapf(
+			types.ErrInvalidBasicMsg,
+			"invalid unlock amount. collateral must more than or equals 0.",
+		)
+	}
+
+	cdp.CollateralAmount = collateralAmount.Uint64()
+
+	// Calculate new collateral ratio. If collateral is lower than 150 percent then returns error.
+	debtAmount := new(big.Int).SetUint64(cdp.DebtAmount)
+	minimumDebtAmount := new(big.Int).SetUint64(0)
+	if debtAmount.Cmp(minimumDebtAmount) > 0 {
+
+		collateralRatioFloat := calculateCollateralRatioOfCDP(cdp, collateralPrice, AtomMultiplier)
+		minimunRatio := new(big.Float).SetFloat64(MinimumCollateralRatio)
+		collateralRatio, _ := collateralRatioFloat.Float64()
+		if collateralRatioFloat.Cmp(minimunRatio) == -1 {
+			return sdkerrors.Wrapf(
+				types.ErrTooLowCollateralRatio,
+				fmt.Sprintf("collateral rate is too low. (%f%)", collateralRatio),
+			)
+		}
+	}
+
+	err = keeper.SupplyKeeper.SendCoinsFromModuleToAccount(ctx, ModuleName, msg.Sender, unlockAmountCoins)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, "can't transfer coins from CDP module to sender")
+	}
+
+	// Store CDP
+	keeper.SetCDP(ctx, cdp)
+
+	return nil
+}
+
+// handleMsgBorrowDebt handles the borrow debt message after receives oracle packet
+func handleMsgBorrowDebt(ctx sdk.Context, keeper Keeper, msg types.MsgBorrowDebt, collateralPrice uint64) error {
+	cdp := keeper.GetCDP(ctx, msg.Sender)
+
+	borrowAmount := sdk.NewCoin(types.MeiUnit, sdk.NewInt(int64(msg.Amount)))
+	borrowAmountCoins := sdk.NewCoins(borrowAmount)
+
+	// Accumurate debt on CDP
+	borrowAmountInt := new(big.Int).SetUint64(msg.Amount)
```

Comment:
We should convert this to `sdk.Coin` or `sdk.Coins` and use comparison function from there. Repeat this everywhere.

Reply:
Converted to coin type